### ISSUE TITLE
Have to reconnect external storage to drawer controller every time world loads

### ISSUE THREAD

**Issue by onlyamonth** [refinedmods/refinedstorage#2636]
Issue description:
Have to reconnect external storage to drawer controller every time world loads

What happens:
Basically the same thing as #2120, but in 1.15.2

What you expected to happen:
Storage is seen on world load

Steps to reproduce:
Make a drawer controller and some drawers, connect to external storage
Put stuff in drawers, appears in grid
Reload world - drawer contents do not appear in grid
...
Version (make sure you are on the latest version before reporting):
Minecraft: 1.15.2
Forge: 31.2.21
Refined Storage: 1.8.8
Does this issue occur on a server? yes

Modpack: Valhelsia2 2.2.3, with latest versions of refined storage, addons and storage drawers.

**Comment by raoulvdberge** [comment]
Can you interact with the drawer controller with hoppers etc after world reload?

**Comment by onlyamonth** [comment]
Yes, and that showed an interesting behaviour - the external bus sees items added to connected drawers after a new world load, but never gets the pre-existing item amounts. It's like it see's all the drawers, knows what goes in them, but thinks they're all empty. 

It inserts into the drawers when new items are added via grid. You can extract those items again, but not any more than what you put in since world load, even if there are more in the drawers. Replacing the storage bus makes all existing items appear.

So, it seems on load it thinks the drawers are empty but knows what should go in them.

**Comment by MysticDaedra** [comment]
I have this same issue, currently playing the MC Eternal modpack (not sure if that matters or not). The issue is specifically External Storages and Drawer Controllers from Storage Drawers mod. My temporary fix is to break and replace the external storages after loading the world, but if I forget a lot of my machines will start to fill up my drive storage.

**Comment by allout58** [comment]
I have a similar issue in the AllTheMods6 pack. I've found if I just cycle through the import/export/both modes after startup, it solves the issue

**Comment by tivec** [comment]
Just to chime in, I'm seeing the same issue.

**Comment by joseph-p-pasaoa** [comment]
Same exact issue as @onlyamonth and others as above. I'm playing:
FTB Revelation (MC 1.12.2, Forge 14.23.5.2846)
Storage Drawers 1.12.2-5.4.2

Originally, my personal half-solution is a relay (off on rs signal) enroute to the Ext Storage with a button that have to press after every world start otherwise as stated by others the containers are seen but read as empty. Items that my mob farm feeds into the drawers do actually show up after world start but the items that were in the drawers at world load are not visible until the inventory is re-checked like in my half-solution above.

I've been playing with the issue for the past hour reorienting the drawer controller, placing it in different locations, having the external storage connect to different faces but nothing has solved the issue yet. I isolated my refined storage to a RS Controller, Grid, and External Storage and the issue still happens. The drawers block as well as my RS system are all contained within one chunkloaded chunk. I've tried substituting Phantomfaces proxying the Drawer Controller connected to the Ext Storage instead, still fails.

I hope any of this helps. Let me know if there's anything I can do to help fig this out.

**Comment by embeddedt** [comment]
> Originally, my personal half-solution is a relay (off on rs signal) enroute to the Ext Storage with a button that have to press after every world start otherwise as stated by others the containers are seen but read as empty.

This is a smart workaround.

(As far as I know, 1.12 is not supported anymore by RS so this will never be fixed upstream in that version. However, https://github.com/tth05/morerefinedstorage might be willing to accept a fix if one is found.)

**Comment by kamild1996** [comment]
Confirmed it's still an issue on MC 1.16.4, All The Mods 6 1.3.7 (Refined Storage 1.9.9, Storage Drawers 8.2.1).

**Comment by embeddedt** [comment]
Can someone provide a world (ideally with only the most necessary mods installed) which can be used to reproduce this? I can't get this to happen with just RS and Storage Drawers.

**Comment by joseph-p-pasaoa** [comment]
I've discovered the chunk with the connection also needs to be chunk loaded using the FTB chunk loader. Any other chunk loaded like chicken chunks won't replicate the issue

**Comment by embeddedt** [comment]
@joseph-p-pasaoa I've tried chunkloading the entire RS system and just the external storage/storage drawer part using FTB Utilities, however, the issue is still not happening. Is there anything else special about your system that might affect how chunk reloading works?

> Any other chunkloader like chicken chunks won't replicate the issue

It would be interesting to see whether this is also true in the post-1.13 versions (as I believe chunkloading mechanics changed).

**Comment by V8gaming** [comment]
I believe it happens with a large storage drawer, it works with like 1 & 8 but when its like 40 4x4 drawers it doesn't seem to load, maybe its a overload of some-sort.

**Comment by embeddedt** [comment]
@V8gaming Thanks for that tip; that means that the issue may be related to the storage drawer system overflowing into adjacent chunks. I'll experiment with a larger drawer setup.

**Comment by LunardH** [comment]
> I've discovered the chunk with the connection also needs to be chunk loaded using the FTB chunk loader. Any other chunkloader like chicken chunks won't replicate the issue

I was chunk loading with chicken chunks and had the issue, removed my chunk loader and still have it, will try removing reborn storage as someone said it fixed for him but idk

**Comment by embeddedt** [comment]
I've experimented with some different setups and still haven't been able to replicate this. It would be very useful if someone could send a world which has the problem, otherwise I'm not sure I or anyone else will be able to fix the issue.

**Comment by LunardH** [comment]
> I've experimented with some different setups and still haven't been able to replicate this. It would be very useful if someone could send a world which has the problem, otherwise I'm not sure I or anyone else will be able to fix the issue.

Can send u the mods and config folder and then my world  for you to see, for now ive done the redstone signal thing but this gonna be crap whenever im not at the place

**Comment by embeddedt** [comment]
That'd be great!

**Comment by LunardH** [comment]
Here it is https://drive.google.com/file/d/13d4BJmLTHVEIfpaIo9A_ECBlhGFTbvNN/view?usp=sharing

**Comment by LunardH** [comment]
And Reborn Storage was not the problem in my case

**Comment by LunardH** [comment]
> Here it is https://drive.google.com/file/d/13d4BJmLTHVEIfpaIo9A_ECBlhGFTbvNN/view?usp=sharing

oh shit, imma need to tell u the coords but i dont have them rn and gonna be out for an hour i think

**Comment by Wishbone1977** [comment]
Okay, so I've done a bit of testing, and have gotten some results.

I took a copy of my pack to mess around with, and tried running some tests with my existing world. After each change I made, I did a "Save & Quit" and then started the world again to see if the problem persisted.

1. Uninstall FTB utilities altogether.
_Problem persists._
2. Break and replace both the drawer controller and the external storage.
_Problem persists._
3. Break both the drawer controller and the external storage and move them to the same chunk as the Refined Storage controller.
_Problem goes away._
4. Break both the drawer controller and the external storage and move them back to where they were (neighboring chunk).
_Problem returns._

In other words, it looks very much like the problem is a cross-chunk issue of some kind. My storage controller is right up against a chunk border, and my wall of drawers is up against the same border, but on the other side of the border. My drawer controller was on the other side of the drawer wall, which is why I could simply move it to the near side of the wall instead and have it end up in the same chunk as the storage controller.

In all cases where the problem manifested itself, I could make it go away by flicking a lever placed on the block next to the external storage. It would then remain absent until the world was next reloaded.

Just to be absolutely sure, I went back to my original world (that still had FTB Utils installed), restored all the chunkloading I had before, and ran tests 2-4 again, with exactly the same results. While that doesn't rule out other mods as a possible source of the problem, FTB Utilities is certainly not to blame.

Let me know if there is anything else you would like me to test.

**Comment by LunardH** [comment]
@Wishbone1977 In your own world did the same chunk thing resolve the issue? Imma test it right now

**Comment by LunardH** [comment]
But anyways its kinda crap cause ur gonna have to put every external storage in this said chunk

**Comment by LunardH** [comment]
Jus tested and it did not work for me putting them in the same chunk

**Comment by Wishbone1977** [comment]
> @Wishbone1977 In your own world did the same chunk thing resolve the issue? Imma test it right now

Yes, as I said, I got the same results from the tests in my own world (well, same world really, but my original copy of it which included chunk loading with FTB Utilities). In other words, when moving the drawer controller and external storage into the same chunk as the storage controller, the problem disappears, and the materials in the drawer system are available in the RS grid right after loading the world. When moving them back into the neighboring chunk, the materials are missing from the grid after loading the world, until a block update is forced on a block connected to the external storage (flicking a lever in my case).

**Comment by Wishbone1977** [comment]
> But anyways its kinda crap cause ur gonna have to put every external storage in this said chunk

Hehe, I didn't say it was a _solution_ to the problem. It's just an indication of what might be a _cause_ of the problem. I'm just playing detective here.

**Comment by LunardH** [comment]
I know wasnt talking about u when i said crap, this bug is crap

**Comment by tivec** [comment]
Any progress on locating the source of this bug? While you can get around it with a redstone signal, it's really hampering the enjoyability of the mod for people who play on local machines..

**Comment by LunardH** [comment]
@tivec The problem solved itself for me, idk what i did or if I even did anything, but started to have issues with the multiblock from reborn kinda the same way, when I'd put any new patterns it wasnt updating on the grid, solved it by simply putting a timer conected to the multiblock giving it a block update every half of a second almost

**Comment by LunardH** [comment]
Maybe just a block update fixes it, try a timer in the block that is the last connection before the external storage

**Comment by MarkSzita** [comment]
> Same exact issue as @onlyamonth and others as above. I'm playing:
> FTB Revelation (MC 1.12.2, Forge 14.23.5.2846) SINGLE PLAYER
> Refined Storage 1.6.15
> Storage Drawers 1.12.2-5.4.2
> RebornStorage 3.3.4.1
> Refined Storage Addons 0.4.4
> 
> Originally, my personal half-solution is a relay (off on rs signal) enroute to the Ext Storage with a button that have to press after every world start otherwise as stated by others the containers are seen but read as empty. Items that my mob farm feeds into the drawers do actually show up after world start but the items that were in the drawers at world load are not visible until the inventory is re-checked like in my half-solution above.
> 
> I've been playing with the issue for the past hour reorienting the drawer controller, placing it in different locations, having the external storage connect to different faces but nothing has solved the issue yet. I isolated my refined storage to a RS Controller, Grid, and External Storage and the issue still happens. The drawers block as well as my RS system are all contained within one chunkloaded chunk. I've tried substituting Phantomfaces proxying the Drawer Controller connected to the Ext Storage instead, still fails.
> 
> I hope any of this helps. Let me know if there's anything I can do to help fig this out.

A very useful workaround. I have expanded upon it for wireless usage, as i often logout away from the main part of my base.
rename an item starting with 1 so that it shows up on the grid right in the first spot (assuming you sort by name ascending).
then use a detector to detect just one of said item. it will put out a redstone signal that you can use to toggle the external storage just by taking the stick out then putting it back into the storage via your favourite wireless grid.

**Comment by Goldy93** [comment]
So, my system stopped having the issue and the only thing I did was put a storage monitor on top of my disk drive and that storage monitor happens to be looking at an item in my storage drawers.  Now, I am not sure if it matters where I placed the monitor (on the disk drive) or that it is looking for something located in my storage drawers.

**Comment by LunardH** [comment]
@Goldy93 The problem will stop if you have any block update in the external storage thats connected to the drawer, like a timer or something else

**Comment by DizzyDJW** [comment]
It sounds like the mod author needs to just add a block update ability to the External Storage block, that detects when a player loads into the world then updates the External Storage block twice, the second update for redundancy purposes.

**Comment by tivec** [comment]
> It sounds like the mod author needs to just add a block update ability to the External Storage block, that detects when a player loads into the world then updates the External Storage block twice, the second update for redundancy purposes.

Maybe setting the block update to a few moments after the world has loaded, to avoid issues with load order etc. I haven't got the skill to make a pull request for this, since I unfortunately lack experience with Forge...

**Comment by embeddedt** [comment]
Has anyone gotten this to happen with chests or is it limited to storage drawer controllers only?

**Comment by tivec** [comment]
> Has anyone gotten this to happen with chests or is it limited to storage drawer controllers only?

It is limited to a few storage blocks, StorageDrawer controllers are most common since they're something people use lots. The drawers themselves seem to work, so if you can put a storage bus on each...

**Comment by DizzyDJW** [comment]
There used to be a mod that added a block that could detect a specific player in a server and emit a redstone signal based on whether that player was there or not. Using that block you could just set it next to the External Storage and let it update the block for you.

**Comment by tivec** [comment]
> There used to be a mod that added a block that could detect a specific player in a server and emit a redstone signal based on whether that player was there or not. Using that block you could just set it next to the External Storage and let it update the block for you.

We have workarounds galore, but no fix for the problem. I'd rather have a fix to the bug.

**Comment by embeddedt** [comment]
Part of the problem with fixing this bug is that it doesn't seem to be reproducible in fresh worlds with small storage systems. I have tried a number of times over the last few months to reproduce this and haven't been successful.

Does anyone have a world in a relatively small modpack (less than 20 mods) that has this issue? Worlds from the big modpacks like ATM6 are unfortunately not that helpful for two reasons:

* They take quite a long time to load.
* It is usually very difficult to get all of the mods into a Forge development environment.

**Comment by Darkere** [comment]
If you attach a debugger to the running pack instead of loading completely from IDE (you need some special args for it) you can just debug RS + Storagedrawers. 
(although performance is likely pretty bad).

**Comment by tivec** [comment]
> Part of the problem with fixing this bug is that it doesn't seem to be reproducible in fresh worlds with small storage systems. I have tried a number of times over the last few months to reproduce this and haven't been successful.
> 
> 
> 
> Does anyone have a world in a relatively small modpack (less than 20 mods) that has this issue? Worlds from the big modpacks like ATM6 are unfortunately not that helpful for two reasons:
> 
> 
> 
> * They take quite a long time to load.
> 
> * It is usually very difficult to get all of the mods into a Forge development environment.

I'm relatively certain it has been confirmed to happen when the storage controller and bus is in a chunk different from the RS controller, and it has been reproduced in smaller mod setups. 

As said by the next post over, a debugger can be attached but it is still very tricky to track it down. I would help code wise, but again I have no dev environment nor forge dev experience.

**Comment by embeddedt** [comment]
Indeed, the number of reports alone confirms that it must be an issue, but there must be more to it than just the drawer controller and external storage bus in a separate chunk from the controller. Here is an example of one of the setups I tried in 1.12. I believe I tried this in 1.16 as well and got the same results.

![image](https://user-images.githubusercontent.com/42941056/126910351-6c7d21e9-b300-4c2c-a688-bb2187377502.png)

Logging out and back in still displays the drawer items in the crafting grid.

**Comment by Darkere** [comment]
You might need to chunkload it using FTBChunks to get the error to occur.

**Comment by embeddedt** [comment]
I'm now trying in 1.16 but still can't get the exact behavior described in this issue to occur reliably.

What I have figured out is this: the controller does *not* attempt to keep loading the chunk the external storage interface and drawers are in. In a test setup where the controller and external storage were placed more than 1 render distance away from each other, I logged out while standing next to the controller, and when logging back in, the items in the drawers were no longer listed.

However, once I load the drawer controller and external storage interface with FTB Chunks, this issue goes away, and the system works normally, even after relogging.

**Comment by DizzyDJW** [comment]
My RS and Storage Drawers have ALWAYS been in the same chunk, to save on resources when keeping chunks loaded, so I can say without a doubt that having them in separate chunks is not the culprit.

However FTBChunks may somehow be affected it negatively via the load order in which the chunks update, potentially, slower than RS blocks do. Just a theory, but I've never not had a chunk loader with my RS and SD setup.

**Comment by embeddedt** [comment]
I cannot get anything amiss to happen when everything is in one chunk. I'm convinced that the problem exists, but it refuses to appear in my test world.

**Comment by Sakata-MC** [comment]
Still ongoing in my ATM6 1.7.6 world. Not sure what the cause is, but just takes a break/replace fo the cable and it works until the server restarts.

**Comment by Wishbone1977** [comment]
> You might need to chunkload it using FTBChunks to get the error to occur.

Nope, I replicated it consistently _without_ FTB Utils installed: https://github.com/refinedmods/refinedstorage/issues/2636#issuecomment-754737614

**Comment by embeddedt** [comment]
> Oddly doesn't occur with Dank storages, so may be something specifically with RS and SD.

I am wondering if the storage drawer controller takes an extra tick to actually display the right contents, but doesn't notify RS when it has updated its contents on that tick. That's just a guess, though.

**Comment by NiSiSuinegEht** [comment]
I am experiencing this issue in Enigmatica E6 0.5.11.  Entire drawer system and RS system are in the same FTBChunks loaded chunk.  Have to break and re place external storage after server restarts.

**Comment by Sakata-MC** [comment]
A lame workaround if computercraft is installed:
Put a computer next to the RS external storage. Set the RS storage to only
work *without* a redstone signal.
On the computer, run this command (assuming you have wget):
wget
https://raw.githubusercontent.com/Sakata-MC/computercraft/main/RefinedStorage_StorageDrawers_StorageBus_Fix.lua
startup

edit the script: edit startup

make sure the variable is set to the correct side, reboot it and make sure
it doesn't explode :D

**Comment by tivec** [comment]
Feels like this could probably be solved if RS waited a second or so to allow the chunks to load before checking storage buses, but that might not work over chunk borders...

I dunno, I will use the workarounds for now. Maybe the multi block from the drawer controller is loaded later and that makes RS fail to see it. 

If storage buses have a status if they are connected or not, maybe it would be possible to check if they aren't and then read the inventory?

**Comment by NiSiSuinegEht** [comment]
I feel the simple solution would be for the external storage to trigger a bock update whenever it is queried by the controller.

**Comment by raoulvdberge** [comment]
Hi guys, this will be solved in the next version for RS (v1.11.3) on MC 1.19.2.

The fix will also be backported to 1.18.2 and will come to CurseForge soon.
